File: crates/sta/src/ext_shim.js
// The "current tab" shim sta evaluates in an extension's popup page and service worker while that
// extension's popup card is open (ext_shim.rs; docs/research/extensions.md §6).
//
// Chromium's `tabs.query` / `windows.*` walk Chrome `Browser` windows, and sta's tabs are Alloy views
// in none of them, so "the current tab" is `[]` / "No current window" for every extension. What
// *does* work natively is `tabs.get(<id>)` for an sta tab. This shim therefore only supplies the
// one thing Chromium cannot know — which tab the popup was opened over — and asks the extension's
// own `chrome.tabs.get` for everything else, so what an extension may see of that tab (its URL,
// its title) is still decided by Chromium from the extension's own permissions.
//
// It is an expression: `(<this file>)(<config>)`, with
//   { popupUrl, url, delta, tabId? }   in the popup page (it finds the tab id itself, see `resolve`)
//   { popupUrl, url, tabId?, near? }   in the service worker (`tabId` is what a page found)
//   { popupUrl: null, url: null }      in an extension page that is itself an sta tab: no tab is
//                                      named (it may be in the background), only `tabs.create`
//                                      gets its fallback
// and it answers 'installed' | 'updated' | 'no-extension-api'.
(cfg) => {
  'use strict';
  const api = typeof chrome === 'object' && chrome ? chrome : null;
  if (!api || !api.tabs || !api.runtime || !api.runtime.id) return 'no-extension-api';

  const KEY = Symbol.for('sta.currentTab');
  const existing = globalThis[KEY];
  if (existing) {
    // A tab id the page already found stays good for as long as it is about the same tab.
    if (typeof cfg.tabId === 'number' || existing.cfg.url !== cfg.url || existing.cfg.delta !== cfg.delta || existing.cfg.near !== cfg.near) existing.tabId = typeof cfg.tabId === 'number' ? cfg.tabId : undefined;
    existing.cfg = cfg;
    return 'updated';
  }
  const state = (globalThis[KEY] = { cfg, tabId: typeof cfg.tabId === 'number' ? cfg.tabId : undefined });

  // The window sta's tabs pretend to be in. Real window ids are SessionIDs (large), so 1 is free.
  const WINDOW_ID = 1;
  // How many SessionIDs Chromium may have handed to something that is not an sta browser (a
  // Chrome-created window) between the tab and the popup.
  const MAX_SKEW = 16;
  const isPage = typeof document === 'object';

  const tabs = api.tabs;
  // Not every extension context has `chrome.windows`; the tab answers do not depend on it.
  const windows = api.windows || null;
  const WINDOW_ID_CURRENT = windows && typeof windows.WINDOW_ID_CURRENT === 'number' ? windows.WINDOW_ID_CURRENT : -2;
  const native = {
    query: tabs.query.bind(tabs),
    get: tabs.get.bind(tabs),
    getCurrent: tabs.getCurrent ? tabs.getCurrent.bind(tabs) : null,
    create: tabs.create.bind(tabs),
    ...(windows && {
      winGet: windows.get.bind(windows),
      winGetCurrent: windows.getCurrent.bind(windows),
      winGetLastFocused: windows.getLastFocused.bind(windows),
      winGetAll: windows.getAll.bind(windows),
      winCreate: windows.create.bind(windows),
      winUpdate: windows.update.bind(windows),
    }),
  };

  const sameUrl = (a, b) => typeof a === 'string' && typeof b === 'string' && (a === b || a.split('#')[0] === b.split('#')[0]);

  // Is the popup this state belongs to still open? The page *is* the popup. A service worker
  // outlives it, and must stop answering for a tab the user has long left.
  const alive = async () => {
    if (isPage) return true;
    if (!api.runtime.getContexts || typeof state.cfg.popupUrl !== 'string') return false;
    try {
      const contexts = await api.runtime.getContexts({});
      return contexts.some((c) => typeof c.documentUrl === 'string' && c.documentUrl.split(/[?#]/)[0] === state.cfg.popupUrl.split(/[?#]/)[0]);
    } catch {
      return false;
    }
  };

  // Tab ids and CEF browser ids are handed out in the same order, so the tab is `delta` ids below
  // the popup — further when Chromium created windows of its own in between. The URL sta knows
  // the tab by is the proof; an extension that may not read URLs gets no answer (it could not use
  // one: nothing in sta grants `activeTab`).
  //
  // A service worker has no tab of its own to count from. Until the page's answer reaches it, it
  // looks around `near`: where the tab would be if nothing had moved since sta last saw an id.
  const resolve = async () => {
    if (typeof state.tabId === 'number') return state.tabId;
    const cfg = state.cfg;
    const candidates = [];
    if (isPage && native.getCurrent && typeof cfg.delta === 'number') {
      const self = await native.getCurrent();
      if (self && typeof self.id === 'number') for (let skew = 0; skew <= MAX_SKEW; skew++) candidates.push(self.id - cfg.delta - skew);
    } else if (typeof cfg.near === 'number') {
      candidates.push(cfg.near);
      for (let skew = 1; skew <= MAX_SKEW; skew++) candidates.push(cfg.near + skew, cfg.near - skew);
    }
    for (const id of candidates) {
      const tab = await native.get(id).catch(() => null);
      if (tab && sameUrl(tab.url || tab.pendingUrl, cfg.url)) {
        if (state.cfg === cfg) state.tabId = tab.id;
        return tab.id;
      }
    }
    return undefined;
  };
  // ext_shim.rs asks the page for the id, to hand it to the service worker.
  state.resolve = resolve;

  const currentTab = async () => {
    if (!(await alive())) return null;
    const id = await resolve();
    if (typeof id !== 'number') return null;
    const tab = await native.get(id).catch(() => null);
    return tab ? { ...tab, active: true, highlighted: true, selected: true, index: 0, windowId: WINDOW_ID } : null;
  };

  const fakeWindow = (tab, populate) => ({
    id: WINDOW_ID,
    focused: true,
    incognito: false,
    alwaysOnTop: false,
    state: 'normal',
    type: 'normal',
    top: 0,
    left: 0,
    width: tab.width || 1280,
    height: tab.height || 800,
    ...(populate ? { tabs: [tab] } : {}),
  });

  const matchesPattern = (pattern, url) => {
    if (pattern === '<all_urls>') return /^(https?|file|ftp|wss?):/.test(url);
    const escaped = pattern.replace(/[.+?^${}()|[\]\\]/g, '\\$&').replace(/\*/g, '.*');
    return new RegExp(`^${escaped}$`).test(url);
  };

  // A query is about "the current tab" when it names the current window, or the active tab of no
  // particular window. Everything else (`{}`, another window's tabs) stays Chromium's.
  const asksForCurrent = (q) => {
    if (!q || typeof q !== 'object') return false;
    if (q.active === false || q.highlighted === false || q.currentWindow === false || q.lastFocusedWindow === false) return false;
    if (q.pinned === true || q.audible === true || q.muted === true || q.discarded === true || q.frozen === true) return false;
    if (typeof q.index === 'number' && q.index !== 0) return false;
    if (typeof q.groupId === 'number' && q.groupId !== -1) return false;
    if (typeof q.windowType === 'string' && q.windowType !== 'normal') return false;
    if (typeof q.windowId === 'number') return q.windowId === WINDOW_ID || q.windowId === WINDOW_ID_CURRENT;
    return q.active === true || q.currentWindow === true || q.lastFocusedWindow === true;
  };

  const passesFilters = (q, tab) => {
    if (q.url !== undefined) {
      const patterns = Array.isArray(q.url) ? q.url : [q.url];
      if (typeof tab.url !== 'string' || !patterns.some((p) => matchesPattern(String(p), tab.url))) return false;
    }
    if (typeof q.status === 'string' && q.status !== tab.status) return false;
    if (typeof q.title === 'string' && !matchesPattern(q.title, tab.title || '')) return false;
    return true;
  };

  // `ours` answers with a value, or `undefined` for "not mine" — then the native call runs with the
  // caller's own arguments, callback included, so its result and `runtime.lastError` are Chromium's.
  const route = (nativeFn, ours) =>
    function (...args) {
      const callback = typeof args[args.length - 1] === 'function' ? args[args.length - 1] : null;
      const plain = callback ? args.slice(0, -1) : args;
      const answer = Promise.resolve()
        .then(() => ours(...plain))
        .catch(() => undefined);
      if (callback) {
        answer.then((value) => (value === undefined ? nativeFn(...args) : callback(value)));
        return undefined;
      }
      return answer.then((value) => (value === undefined ? nativeFn(...plain) : value));
    };

  tabs.query = route(native.query, async (q) => {
    if (!asksForCurrent(q)) return undefined;
    const tab = await currentTab();
    if (!tab) return undefined;
    return passesFilters(q, tab) ? [tab] : [];
  });

  const currentWindow = async (options) => {
    const tab = await currentTab();
    return tab ? fakeWindow(tab, !!(options && options.populate)) : undefined;
  };
  if (windows) {
    windows.getCurrent = route(native.winGetCurrent, currentWindow);
    windows.getLastFocused = route(native.winGetLastFocused, currentWindow);
    windows.get = route(native.winGet, async (id, options) => (id === WINDOW_ID ? currentWindow(options) : undefined));
    windows.update = route(native.winUpdate, async (id) => (id === WINDOW_ID ? currentWindow() : undefined));
    windows.getAll = route(native.winGetAll, async (options) => {
      const ours = await currentWindow(options);
      if (!ours) return undefined;
      const rest = await native.winGetAll(options).catch(() => []);
      return [ours, ...rest];
    });
  }

  // A profile without a Chrome window cannot open a tab ("No current window"). `windows.create`
  // can: Chromium makes a window of its own, which sta hides and turns into one of its tabs
  // (foreign.rs), under the same verdict and budget as every other tab an extension asks for.
  tabs.create = route(native.create, async (props) => {
    const wanted = props && typeof props === 'object' ? { ...props } : {};
    if (wanted.windowId === WINDOW_ID) delete wanted.windowId;
    if (typeof wanted.windowId === 'number' && wanted.windowId !== WINDOW_ID_CURRENT) return undefined;
    try {
      return await native.create(wanted);
    } catch (e) {
      if (!/No current window/i.test(String(e && e.message))) return undefined;
    }
    if (!native.winCreate) return undefined;
    const created = await native.winCreate({ url: wanted.url, focused: wanted.active !== false, type: 'normal' });
    const tab = created && created.tabs && created.tabs[0];
    return tab ? { ...tab, windowId: WINDOW_ID } : undefined;
  });

  // What pressing the extension's toolbar button would do right now — which is the extension's to
  // decide at run time, not its manifest's: `action.setPopup('')` means "no popup, tell me about
  // the click" (1Password without an account opens its sign-in page that way). sta has no toolbar
  // button, so ext_shim.rs asks the service worker this once per card, and an empty popup gets the
  // click it was waiting for.
  state.action = async () => {
    const action = api.action || api.browserAction;
    if (!action || !action.getPopup || !action.onClicked) return { popup: null, clicked: false };
    const tab = await currentTab();
    let popup;
    try {
      popup = await action.getPopup(tab ? { tabId: tab.id } : {});
    } catch {
      popup = await action.getPopup({}).catch(() => null);
    }
    if (popup !== '') return { popup, clicked: false };
    if (!action.onClicked.hasListeners() || typeof action.onClicked.dispatch !== 'function') return { popup, clicked: false };
    const none = typeof tabs.TAB_ID_NONE === 'number' ? tabs.TAB_ID_NONE : -1;
    action.onClicked.dispatch(tab || { id: none, index: -1, windowId: -1, active: true, highlighted: true, selected: true, pinned: false, incognito: false, discarded: false, autoDiscardable: true, groupId: -1 });
    return { popup, clicked: true };
  };

  // The page's answer is what the service worker waits for: have it ready when sta asks.
  if (isPage) resolve().catch(() => {});
  return 'installed';
}

File: crates/sta/src/ext_shim.rs
//! "The current tab" for an extension whose popup card is open [owner: chrome]
//! (docs/research/extensions.md §6).
//!
//! Chromium's `tabs.query` / `windows.*` only walk Chrome `Browser` windows, and sta's tabs are
//! Alloy views in none of them — so a popup that asks for the tab it was opened over gets `[]`, and
//! its service worker "No current window". What *does* work natively is `tabs.get(<id>)` with an
//! sta tab's id. So while a popup card is open, sta evaluates one script (`ext_shim.js`) in
//! - **the popup page** — registered for its document before the page's own scripts run
//!   (`Page.addScriptToEvaluateOnNewDocument`), and evaluated again when the page has loaded;
//! - **that extension's service worker** — Chromium attaches the popup's DevTools session to
//!   service workers as they start (`Target.setAutoAttach`); a worker of any other extension is let
//!   go of at once.
//!
//! The script only supplies which tab the card was opened over. The tab's id is found **in the
//! popup page**: tab ids and CEF browser ids are handed out in the same order, so the tab is
//! `popup browser id − tab browser id` ids below the popup's own (`tabs.getCurrent`), and the URL
//! sta knows the tab by has to match what the extension's own `tabs.get` answers. That URL is part
//! of the script's configuration, which the extension can read — so it is only handed to an
//! extension whose manifest lets it read that URL anyway (`ExtensionFiles::may_read_url`). One
//! that may not never finds the tab itself, and everything an extension learns about the tab is
//! what Chromium lets *that extension* see. An id a page found is remembered for
//! its browser, and the last pair tells a service worker where to look (`near`) before the page
//! of a tab nobody asked about yet has answered — a popup's first message is often "which tab?".
//!
//! It ends with the popup: the page is gone, the worker session closes with the popup's browser,
//! and the worker's copy checks that the popup is still open before every answer.
//!
//! The same script lets `tabs.create` fall back to `windows.create` when the profile has no Chrome
//! window ("No current window"): that window is one `foreign.rs` hides and turns into an sta tab,
//! under core's usual verdict and budget. An extension page that is itself an sta tab gets the
//! script for that alone ([`on_extension_tab_document`]): its "Sign in" is a `tabs.create` too.
//!
//! **The toolbar click.** What pressing an extension's button does is the extension's to decide
//! at run time, not its manifest's: `action.setPopup('')` means "no popup — tell me about the
//! click" (1Password without an account opens its sign-in page that way, and its popup page, which
//! was never meant to be seen then, stays on the logo). sta has no button to press, but the
//! worker session can deliver the event: once per card the script asks `action.getPopup`, and an
//! empty popup with `onClicked` listeners gets `onClicked.dispatch(<the card's tab>)` — after which
//! the card, which has nothing to show, closes.
//!
//! Public API:
//! - `pub fn on_popup_created(popup_browser: i32, extension: &str, popup_url: &str, tab: Option<Id>)`
//! - `pub fn on_popup_loaded(popup_browser: i32)`, `pub fn on_popup_closed(popup_browser: i32)`
//! - `pub fn on_extension_tab_document(frame: &Frame, url: &str)`
//! - `pub fn clear()`, `pub fn debug_snapshot() -> serde_json::Value`

use crate::devtools_cdp::{self, User, WorkerTarget};
use crate::{controller, tabs, task};
use cef::*;
use serde_json::{Value, json};
use sta_core::{Command, Id};
use std::cell::{Cell, RefCell};
use std::collections::HashMap;

const SHIM_JS: &str = include_str!("ext_shim.js");
const CALL_TIMEOUT_MS: i64 = 3000;
/// The popup page is asked for the tab's id from the moment its browser exists: a service worker
/// that is asked about the current tab by the popup's first message needs the answer by then. So
/// the first second is polled closely (an in-process call each), the rest until [`RESOLVE_FOR_MS`]
/// loosely — a page that has not answered by then cannot see the tab's URL at all.
const RESOLVE_EAGER_MS: i64 = 25;
const RESOLVE_LAZY_MS: i64 = 250;
const RESOLVE_EAGER_FOR_MS: i64 = 1000;
const RESOLVE_FOR_MS: i64 = 4000;
/// Tab ids remembered per browser (one never changes while its browser lives, and neither kind of
/// id is reused); emptied when it gets this large.
const KNOWN_MAX: usize = 512;
/// A service worker is asked what the toolbar button would do once it knows the card's tab — or
/// after this long, for a card over no tab it could be told about.
const ACTION_WAIT_MS: i64 = 600;

struct Shim {
    popup_browser: i32,
    /// `chrome-extension://<id>/`: the only workers that get the script.
    origin: String,
    popup_url: String,
    /// The tab the card was opened over: its browser, and its URL **if this extension may read
    /// it** (`ExtensionFiles::may_read_url`). The script's configuration is the extension's to
    /// read, so a URL Chromium would keep from it must not be in there; without one the tab is
    /// only known when another extension's popup found its id before.
    tab: Option<(Option<String>, i32)>,
    tab_id: Option<i64>,
    /// Worker sessions of this extension, in the order they attached.
    workers: Vec<String>,
    /// The toolbar-click question was asked (once per card).
    action_asked: bool,
}

#[derive(Default, Clone, Copy)]
struct Stats {
    popups: u64,
    resolved: u64,
    unresolved: u64,
    workers: u64,
    strangers: u64,
    clicks: u64,
}

thread_local! {
    static CURRENT: RefCell<Option<Shim>> = const { RefCell::new(None) };
    /// Browser id → the tab id extensions know it by, as popup pages found them.
    static KNOWN: RefCell<HashMap<i32, i64>> = RefCell::new(HashMap::new());
    /// The last pair found: where a tab nobody asked about yet probably is (`near`).
    static LAST: Cell<Option<(i32, i64)>> = const { Cell::new(None) };
    static STATS: Cell<Stats> = Cell::new(Stats::default());
}

fn stats(f: impl FnOnce(&mut Stats)) {
    let mut s = STATS.get();
    f(&mut s);
    STATS.set(s);
}

/// The tab's URL as extensions may be told about it: a page of the web, never one of sta's own.
fn shareable(url: &str) -> bool {
    matches!(sta_core::urls::scheme(url).as_deref(), Some("http" | "https" | "file" | "ftp"))
}

/// `(<ext_shim.js>)(<config>)`, for the page (`tab_id: None`: it finds the id) or a worker.
fn script(shim: &Shim, for_worker: bool) -> String {
    let (url, browser) = match &shim.tab {
        Some((url, browser)) => (json!(url), Some(*browser)),
        None => (Value::Null, None),
    };
    let config = match for_worker {
        // Tab ids and browser ids grow together, so the last pair seen says where this tab is —
        // unless Chromium made windows of its own since, which is why the worker looks *around* it.
        true => {
            let near = browser.zip(LAST.get()).map(|(b, (known_browser, known_tab))| known_tab + i64::from(b - known_browser));
            json!({ "popupUrl": shim.popup_url, "url": url, "tabId": shim.tab_id, "near": near })
        }
        false => json!({ "popupUrl": shim.popup_url, "url": url, "tabId": shim.tab_id, "delta": browser.map(|b| shim.popup_browser - b) }),
    };
    format!("({SHIM_JS})({config})")
}

/// The popup card's browser exists (its page has not committed yet).
pub fn on_popup_created(popup_browser: i32, extension: &str, popup_url: &str, tab: Option<Id>) {
    let tab = tab.and_then(tabs::browser_for_tab).and_then(|b| {
        let url = b.main_frame().map(|f| CefString::from(&f.url()).to_string()).unwrap_or_default();
        if !shareable(&url) || b.identifier() >= popup_browser {
            return None;
        }
        let readable = crate::extension_files::find(extension).is_some_and(|files| files.may_read_url(&url));
        Some((readable.then_some(url), b.identifier()))
    });
    let tab_id = tab.as_ref().and_then(|(_, browser)| KNOWN.with(|k| k.borrow().get(browser).copied()));
    let shim = Shim { popup_browser, origin: format!("chrome-extension://{extension}/"), popup_url: popup_url.to_string(), tab, tab_id, workers: Vec::new(), action_asked: false };
    let source = script(&shim, false);
    CURRENT.with(|c| *c.borrow_mut() = Some(shim));
    stats(|s| s.popups += 1);
    // The Page agent only runs registered scripts while it is enabled.
    devtools_cdp::call(popup_browser, User::Extensions, "Page.enable", json!({}), CALL_TIMEOUT_MS, |_| {});
    devtools_cdp::call(popup_browser, User::Extensions, "Page.addScriptToEvaluateOnNewDocument", json!({ "source": source, "runImmediately": true }), CALL_TIMEOUT_MS, |result| {
        if let Err(e) = result {
            log_warn!("ext_shim: the popup page's script was not registered: {e}");
        }
    });
    devtools_cdp::watch_workers(popup_browser, on_worker_attached);
    let params = json!({ "autoAttach": true, "waitForDebuggerOnStart": false, "flatten": true, "filter": [{ "type": "service_worker", "exclude": false }] });
    devtools_cdp::call(popup_browser, User::Extensions, "Target.setAutoAttach", params, CALL_TIMEOUT_MS, |result| {
        if let Err(e) = result {
            log_warn!("ext_shim: no service worker auto-attach: {e}");
        }
    });
    resolve(popup_browser, 0);
}

/// The popup page's main frame loaded: the one moment it can certainly answer.
pub fn on_popup_loaded(popup_browser: i32) {
    resolve(popup_browser, RESOLVE_FOR_MS);
}

/// Asks the popup page which tab id the card's tab has, again until it knows (`waited` ms so
/// far); from `RESOLVE_FOR_MS` on it is a single question.
fn resolve(popup_browser: i32, waited: i64) {
    // Only a page that may read the tab's URL can prove which tab it is.
    let asks = |s: &&Shim| s.popup_browser == popup_browser && s.tab_id.is_none() && s.tab.as_ref().is_some_and(|(url, _)| url.is_some());
    let source = CURRENT.with(|c| c.borrow().as_ref().filter(asks).map(|s| script(s, false)));
    let Some(source) = source else { return };
    // Installing is idempotent, so a page whose document beat the registered script still gets it.
    let expression = format!("(async () => {{ {source}; const s = globalThis[Symbol.for('sta.currentTab')]; const id = s && s.resolve ? await s.resolve() : undefined; return typeof id === 'number' ? id : null; }})()");
    let params = json!({ "expression": expression, "returnByValue": true, "awaitPromise": true });
    devtools_cdp::call(popup_browser, User::Extensions, "Runtime.evaluate", params, CALL_TIMEOUT_MS, move |result| {
        let id = result.ok().as_ref().and_then(|v| v.get("result")).and_then(|r| r.get("value")).and_then(Value::as_i64);
        match id {
            Some(id) => on_resolved(popup_browser, id),
            None if waited >= RESOLVE_FOR_MS => {}
            None => {
                let delay = if waited < RESOLVE_EAGER_FOR_MS { RESOLVE_EAGER_MS } else { RESOLVE_LAZY_MS };
                if waited + delay >= RESOLVE_FOR_MS {
                    stats(|s| s.unresolved += 1);
                    log_debug!("ext_shim: the popup of browser {popup_browser} cannot see its tab (no URL access, or the tab moved on)");
                    return;
                }
                task::post_ui_delayed(delay, move || resolve(popup_browser, waited + delay));
            }
        }
    });
}

fn on_resolved(popup_browser: i32, tab_id: i64) {
    let workers = CURRENT.with(|c| {
        let mut current = c.borrow_mut();
        let shim = current.as_mut().filter(|s| s.popup_browser == popup_browser && s.tab_id.is_none())?;
        shim.tab_id = Some(tab_id);
        Some((shim.workers.clone(), shim.tab.as_ref().map(|(_, browser)| *browser)))
    });
    let Some((workers, tab_browser)) = workers else { return };
    if let Some(browser) = tab_browser {
        KNOWN.with(|k| {
            let mut known = k.borrow_mut();
            if known.len() >= KNOWN_MAX {
                known.clear();
            }
            known.insert(browser, tab_id);
        });
        LAST.set(Some((browser, tab_id)));
    }
    stats(|s| s.resolved += 1);
    log_debug!("ext_shim: the card's tab is tab {tab_id} to the extension");
    for session in workers {
        install_in_worker(popup_browser, &session);
    }
}

/// Asks the extension's service worker what its toolbar button would do now, once per card: when
/// the worker has the script and knows the card's tab (`forced`: or has waited long enough).
fn ask_action(popup_browser: i32, forced: bool) {
    let session = CURRENT.with(|c| {
        let mut current = c.borrow_mut();
        let shim = current.as_mut().filter(|s| s.popup_browser == popup_browser && !s.action_asked)?;
        let waits_for_tab = shim.tab_id.is_none() && shim.tab.as_ref().is_some_and(|(url, _)| url.is_some());
        if waits_for_tab && !forced {
            return None;
        }
        let session = shim.workers.first()?.clone();
        shim.action_asked = true;
        Some(session)
    });
    let Some(session) = session else { return };
    let expression = "(async () => { const s = globalThis[Symbol.for('sta.currentTab')]; return s && s.action ? JSON.stringify(await s.action()) : null; })()";
    devtools_cdp::worker_evaluate(popup_browser, &session, expression, CALL_TIMEOUT_MS, move |result| {
        let answer = result.ok().as_ref().and_then(|v| v.get("result")).and_then(|r| r.get("value")).and_then(Value::as_str).and_then(|text| serde_json::from_str::<Value>(text).ok());
        let clicked = answer.as_ref().and_then(|a| a.get("clicked")).and_then(Value::as_bool).unwrap_or(false);
        if !clicked {
            return;
        }
        let ours = CURRENT.with(|c| c.borrow().as_ref().is_some_and(|s| s.popup_browser == popup_browser));
        if ours {
            stats(|s| s.clicks += 1);
            log_info!("ext_shim: the extension has no popup right now; its toolbar click was delivered instead");
            controller::dispatch(Command::CloseExtensionPopup);
        }
    });
}

fn on_worker_attached(popup_browser: i32, target: WorkerTarget) {
    let ours = CURRENT.with(|c| {
        let mut current = c.borrow_mut();
        let Some(shim) = current.as_mut().filter(|s| s.popup_browser == popup_browser && target.url.starts_with(&s.origin)) else { return false };
        shim.workers.push(target.session.clone());
        true
    });
    if !ours {
        stats(|s| s.strangers += 1);
        devtools_cdp::worker_detach(popup_browser, &target.session);
        return;
    }
    stats(|s| s.workers += 1);
    install_in_worker(popup_browser, &target.session);
    task::post_ui_delayed(ACTION_WAIT_MS, move || ask_action(popup_browser, true));
}

fn install_in_worker(popup_browser: i32, session: &str) {
    let source = CURRENT.with(|c| c.borrow().as_ref().filter(|s| s.popup_browser == popup_browser).map(|s| script(s, true)));
    let Some(source) = source else { return };
    devtools_cdp::worker_evaluate(popup_browser, session, &source, CALL_TIMEOUT_MS, move |result| match result {
        Ok(_) => ask_action(popup_browser, false),
        Err(e) => log_warn!("ext_shim: the service worker did not take the script: {e}"),
    });
}

/// A main-frame document of a tab: an extension's own page gets the script without a tab to name
/// (the page may be in the background), for the `tabs.create` fallback alone.
pub fn on_extension_tab_document(frame: &Frame, url: &str) {
    if !url.starts_with("chrome-extension://") {
        return;
    }
    let source = format!("({SHIM_JS})({})", json!({ "popupUrl": null, "url": null }));
    frame.execute_java_script(Some(&CefString::from(source.as_str())), Some(&CefString::from("")), 1);
}

/// The popup's browser is gone (sta closed the card, or the page closed itself). Its DevTools
/// session, and with it every worker session, went with it.
pub fn on_popup_closed(popup_browser: i32) {
    CURRENT.with(|c| {
        let mut current = c.borrow_mut();
        if current.as_ref().is_some_and(|s| s.popup_browser == popup_browser) {
            *current = None;
        }
    });
    devtools_cdp::on_browser_closed(popup_browser);
}

pub fn clear() {
    CURRENT.with(|c| *c.borrow_mut() = None);
    KNOWN.with(|k| k.borrow_mut().clear());
    LAST.set(None);
}

#[cfg_attr(not(debug_assertions), allow(dead_code))] // debug.rs only
pub fn debug_snapshot() -> Value {
    let s = STATS.get();
    let current = CURRENT.with(|c| {
        c.borrow().as_ref().map(|shim| {
            json!({
                "popupBrowser": shim.popup_browser,
                "origin": shim.origin,
                "tab": shim.tab.as_ref().map(|(url, browser)| json!({ "url": url, "browser": browser })),
                "tabId": shim.tab_id,
                "workers": shim.workers.len(),
            })
        })
    });
    json!({
        "current": current,
        "known": KNOWN.with(|k| k.borrow().len()),
        "stats": { "popups": s.popups, "resolved": s.resolved, "unresolved": s.unresolved, "workers": s.workers, "strangers": s.strangers, "clicks": s.clicks },
    })
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn only_web_pages_are_shared_with_an_extension() {
        for url in ["https://example.com/login", "http://127.0.0.1:8080/", "file:///C:/page.html"] {
            assert!(shareable(url), "{url}");
        }
        for url in ["sta://settings/", "chrome://extensions/", "devtools://devtools/x", "about:blank", "javascript:1", ""] {
            assert!(!shareable(url), "{url}");
        }
    }

    #[test]
    fn the_page_finds_the_id_and_the_worker_is_given_it() {
        let mut shim = Shim { popup_browser: 9, origin: "chrome-extension://abc/".into(), popup_url: "chrome-extension://abc/popup.html".into(), tab: Some((Some("https://example.com/\"x".into()), 6)), tab_id: None, workers: Vec::new(), action_asked: false };
        // The script is an expression applied to its configuration: `(<file>)(<config>)`.
        let config = |script: String| script.rsplit_once(")(").map(|(_, c)| c.to_string()).unwrap_or_default();
        let page = config(script(&shim, false));
        assert!(page.contains(r#""delta":3"#) && page.contains(r#""tabId":null"#), "{page}");
        assert!(page.contains(r#"https://example.com/\"x"#), "the URL is JSON, not pasted: {page}");
        // A tab nobody asked about yet is expected where the last known pair says it is.
        LAST.set(Some((4, 687099968)));
        assert!(config(script(&shim, true)).contains(r#""near":687099970"#));
        LAST.set(None);
        shim.tab_id = Some(687099970);
        let worker = config(script(&shim, true));
        assert!(worker.contains(r#""tabId":687099970"#) && worker.contains(r#""near":null"#) && !worker.contains("delta"), "{worker}");
        // An extension that may not read the URL is not handed it: the id alone, if one is known.
        shim.tab = Some((None, 6));
        let blind = config(script(&shim, false));
        assert!(blind.contains(r#""url":null"#) && blind.contains(r#""tabId":687099970"#) && !blind.contains("example.com"), "{blind}");
        shim.tab = None;
        assert!(config(script(&shim, false)).contains(r#""url":null"#));
    }
}
